File: test/core/push/fcm_incident_payload_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:critalarm/core/push/fcm_incident_payload.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  final fixture =
      jsonDecode(
            File(
              'test/fixtures/android_delivery_cases.json',
            ).readAsStringSync(),
          )
          as Map<String, dynamic>;

  for (final vector in fixture['payloads'] as List<dynamic>) {
    final testCase = vector as Map<String, dynamic>;
    test(testCase['name'] as String, () {
      final data = (testCase['data'] as Map<String, dynamic>)
          .cast<String, String>();
      final parsed = FcmIncidentPayload.tryParse(data);
      expect(parsed != null, testCase['valid']);
    });
  }
}

File: android/app/src/test/kotlin/app/critalarm/push/FcmIncidentPayloadTest.kt
package app.critalarm.push

import com.google.gson.JsonParser
import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Test

class FcmIncidentPayloadTest {
    @Test
    fun `shared payload vectors parse identically`() {
        val root = JsonParser.parseString(fixture().readText()).asJsonObject
        root.getAsJsonArray("payloads").forEach { element ->
            val vector = element.asJsonObject
            val data = vector.getAsJsonObject("data").entrySet().associate { it.key to it.value.asString }
            val parsed = FcmIncidentPayload.fromData(data)
            assertEquals(vector.get("name").asString, vector.get("valid").asBoolean, parsed != null)
        }
    }

    private fun fixture() = sequenceOf(
        File("test/fixtures/android_delivery_cases.json"),
        File("../../test/fixtures/android_delivery_cases.json"),
    ).first { it.exists() }
}
